File: UnityMcpBridge/Editor/Tools/TriggerDomainReload.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEditor.Compilation;
using UnityEngine;
using Debug = UnityEngine.Debug;

namespace UnityMcpBridge.Editor.Tools
{
    public static class TriggerDomainReload
    {
        // Session ID that survives domain reloads
        private static readonly string SESSION_ID;
        private const string SESSION_ID_KEY = "UnityMcpBridge.SessionId";
        private const string JOBS_KEY_PREFIX = "UnityMcpBridge.DomainReloadJob.";
        private const string ACTIVE_JOBS_KEY = "UnityMcpBridge.ActiveJobIds";
        
        // In-memory state (lost on domain reload)
        private static readonly object lockObj = new object();
        private static Dictionary<string, DomainReloadJob> activeJobs = new Dictionary<string, DomainReloadJob>();
        private static bool handlersRegistered = false;
        private static bool wasRestoredAfterReload = false;
        
        // Static constructor - runs when domain loads
        static TriggerDomainReload()
        {
            // Check if we have an existing session ID (we're being reloaded)
            string existingSessionId = SessionState.GetString(SESSION_ID_KEY, null);
            
            if (!string.IsNullOrEmpty(existingSessionId))
            {
                // We've been reloaded - restore state
                SESSION_ID = existingSessionId;
                wasRestoredAfterReload = true;
                Debug.Log($"[TriggerDomainReload] Restored after domain reload. Session: {SESSION_ID}");
                RestoreJobsFromSessionState();
            }
            else
            {
                // Fresh start - create new session
                SESSION_ID = Guid.NewGuid().ToString();
                SessionState.SetString(SESSION_ID_KEY, SESSION_ID);
                Debug.Log($"[TriggerDomainReload] Fresh start. New session: {SESSION_ID}");
            }
            
            // Always register for updates to process jobs
            EditorApplication.update += ProcessJobs;
        }
        
        [Serializable]
        private class DomainReloadJob
        {
            public string jobId;
            public string action;
            public bool includeLogs;
            public string logLevel;
            public float timeout;
            public double startTime;
            public string status = "running"; // running, completed, failed, timeout
            public bool compilationSucceeded = true;
            public List<string> compilationLogs = new List<string>();
            public List<CompilationError> compilationErrors = new List<CompilationError>();
            public string message = "";
            public bool wasCompiling;
            public bool hadToRefresh;
            
            // Track if compilation events occurred
            public bool compilationStarted = false;
            public bool compilationFinished = false;
        }
        
        [Serializable]
        private class CompilationError
        {
            public string type;
            public string message;
            public string file;
            public int line;
            public int column;
        }
        
        public static object HandleCommand(JObject parameters)
        {
            try
            {
                string action = parameters["action"]?.ToString() ?? "compile_and_reload";
                bool waitForCompletion = parameters["waitForCompletion"]?.ToObject<bool>() ?? true;
                bool includeLogs = parameters["includeLogs"]?.ToObject<bool>() ?? true;
                string logLevel = parameters["logLevel"]?.ToString() ?? "all";
                float timeout = parameters["timeout"]?.ToObject<float>() ?? 300f;
                
                // Special handling for status polling
                if (action == "get_status")
                {
                    string jobId = parameters["jobId"]?.ToString();
                    if (string.IsNullOrEmpty(jobId))
                    {
                        return new
                        {
                            success = false,
                            error = "Missing jobId parameter for get_status action"
                        };
                    }
                    
                    return GetJobStatus(jobId);
                }

                // Create a new job
                string newJobId = Guid.NewGuid().ToString();
                var job = new DomainReloadJob
                {
                    jobId = newJobId,
                    action = action,
                    includeLogs = includeLogs,
                    logLevel = logLevel,
                    timeout = timeout,
                    startTime = EditorApplication.timeSinceStartup,
                    wasCompiling = EditorApplication.isCompiling
                };

                // Store job in memory and persistent storage
                lock (lockObj)
                {
                    activeJobs[newJobId] = job;
                    SaveActiveJobIds();
                }
                SaveJobToSessionState(job);
                
                // Register handlers if needed
                if (!handlersRegistered && includeLogs)
                {
                    RegisterCompilationHandlers();
                    handlersRegistered = true;
                }

                // ALWAYS require confirmation for asset-heavy operations (security: cannot be overridden via TCP)
                bool needsConfirmation = action == "refresh_assets" || action == "compile_and_reload";
                
                if (needsConfirmation)
                {
                    string dialogTitle = "Confirm Asset Refresh";
                    string dialogMessage = action == "refresh_assets" 
                        ? "Asset refresh can take a long time for large projects.\n\nDo you want to proceed with refreshing all assets?"
                        : "Full compile and reload includes asset refresh which can take a long time for large projects.\n\nDo you want to proceed?";
                    
                    bool userConfirmed = EditorUtility.DisplayDialog(
                        dialogTitle,
                        dialogMessage,
                        "Yes, Proceed",
                        "Cancel"
                    );
                    
                    if (!userConfirmed)
                    {
                        // User cancelled - clean up and return
                        lock (lockObj)
                        {
                            activeJobs.Remove(newJobId);
                            SaveActiveJobIds();
                        }
                        
                        return new
                        {
                            success = false,
                            error = "User cancelled the operation",
                            message = "Asset refresh was cancelled by user"
                        };
                    }
                }
                
                // Perform the requested action
                Debug.Log($"[TriggerDomainReload] Starting job {newJobId} - Action: {action}");
                
                switch (action)
                {
                    case "refresh_assets":
                        AssetDatabase.Refresh();
                        job.hadToRefresh = true;
                        Debug.Log("Asset database refresh initiated.");
                        break;
                        
                    case "compile_scripts":
                        CompilationPipeline.RequestScriptCompilation();
                        Debug.Log("Script compilation requested.");
                        break;
                        
                    case "domain_reload":
                        EditorUtility.RequestScriptReload();
                        Debug.Log("Domain reload requested.");
                        break;
                        
                    case "compile_and_reload":
                    default:
                        AssetDatabase.Refresh();
                        job.hadToRefresh = true;
                        CompilationPipeline.RequestScriptCompilation();
                        EditorUtility.RequestScriptReload();
                        Debug.Log("Full compile and reload initiated.");
                        break;
                }
                
                // Save job state after action
                SaveJobToSessionState(job);

                // Return immediately with job ID
                return new
                {
                    success = true,
                    data = new
                    {
                        jobId = newJobId,
                        status = "initiated",
                        message = $"Domain reload job '{newJobId}' started. Poll with action='get_status' and jobId='{newJobId}'",
                        sessionId = SESSION_ID,
                        wasCompiling = job.wasCompiling,
                        hadToRefresh = job.hadToRefresh
                    }
                };
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error in TriggerDomainReload: {ex.Message}\n{ex.StackTrace}");
                return new
                {
                    success = false,
                    error = ex.Message,
                    message = "Failed to trigger domain reload"
                };
            }
        }
        
        private static object GetJobStatus(string jobId)
        {
            DomainReloadJob job = null;
            
            // Try to get from memory first
            lock (lockObj)
            {
                activeJobs.TryGetValue(jobId, out job);
            }
            
            // If not in memory, try to restore from session state
            if (job == null)
            {
                string jobJson = SessionState.GetString(JOBS_KEY_PREFIX + jobId, null);
                if (!string.IsNullOrEmpty(jobJson))
                {
                    try
                    {
                        job = JsonConvert.DeserializeObject<DomainReloadJob>(jobJson);
                        
                        // Add it back to active jobs if it's still running
                        if (job != null && job.status == "running")
                        {
                            lock (lockObj)
                            {
                                activeJobs[jobId] = job;
                                SaveActiveJobIds();
                            }
                        }
                    }
                    catch (Exception ex)
                    {
                        Debug.LogError($"Failed to deserialize job {jobId}: {ex.Message}");
                    }
                }
            }
            
            if (job == null)
            {
                return new
                {
                    success = false,
                    error = $"Job '{jobId}' not found. It may have expired or been cleaned up."
                };
            }
            
            // Build response
            var responseData = new Dictionary<string, object>
            {
                ["jobId"] = job.jobId,
                ["status"] = job.status,
                ["compilationSucceeded"] = job.compilationSucceeded,
                ["duration"] = job.status == "running" 
                    ? EditorApplication.timeSinceStartup - job.startTime 
                    : EditorApplication.timeSinceStartup - job.startTime,
                ["wasCompiling"] = job.wasCompiling,
                ["hadToRefresh"] = job.hadToRefresh,
                ["message"] = job.message,
                ["sessionId"] = SESSION_ID
            };
            
            // Add compilation errors if any
            if (job.compilationErrors.Count > 0)
            {
                responseData["compilationErrors"] = job.compilationErrors;
            }
            
            // Add logs if requested and available
            if (job.includeLogs && job.compilationLogs.Count > 0)
            {
                var filteredLogs = FilterLogs(job.compilationLogs, job.logLevel);
                if (filteredLogs.Count > 0)
                {
                    responseData["logs"] = filteredLogs;
                }
            }
            
            // Clean up completed jobs after they've been retrieved
            if (job.status != "running")
            {
                lock (lockObj)
                {
                    activeJobs.Remove(jobId);
                    SaveActiveJobIds();
                }
                // Keep in SessionState for a while in case of re-polling
            }
            
            return new
            {
                success = true,
                data = responseData
            };
        }
        
        private static void ProcessJobs()
        {
            List<DomainReloadJob> jobsToUpdate = new List<DomainReloadJob>();
            
            lock (lockObj)
            {
                foreach (var job in activeJobs.Values)
                {
                    if (job.status == "running")
                    {
                        jobsToUpdate.Add(job);
                    }
                }
            }
            
            foreach (var job in jobsToUpdate)
            {
                // Check if compilation is complete
                if (!EditorApplication.isCompiling && !EditorApplication.isUpdating)
                {
                    // Check if we were waiting for compilation to start
                    if (job.action != "refresh_assets" && !job.compilationStarted)
                    {
                        // Give it a bit more time for compilation to start
                        var elapsed = EditorApplication.timeSinceStartup - job.startTime;
                        if (elapsed < 2.0) // Wait up to 2 seconds for compilation to start
                        {
                            continue;
                        }
                    }
                    
                    // Compilation is done (or never started)
                    job.status = job.compilationSucceeded ? "completed" : "failed";
                    job.message = job.compilationSucceeded 
                        ? "Domain reload completed successfully" 
                        : "Compilation failed with errors";
                    
                    Debug.Log($"[TriggerDomainReload] Job {job.jobId} completed with status: {job.status}");
                    SaveJobToSessionState(job);
                }
                else
                {
                    // Still compiling - check timeout
                    var elapsed = EditorApplication.timeSinceStartup - job.startTime;
                    if (elapsed > job.timeout)
                    {
                        job.status = "timeout";
                        job.message = $"Operation timed out after {job.timeout} seconds";
                        Debug.Log($"[TriggerDomainReload] Job {job.jobId} timed out");
                        SaveJobToSessionState(job);
                    }
                }
            }
        }
        
        private static void SaveJobToSessionState(DomainReloadJob job)
        {
            try
            {
                string jobJson = JsonConvert.SerializeObject(job);
                SessionState.SetString(JOBS_KEY_PREFIX + job.jobId, jobJson);
            }
            catch (Exception ex)
            {
                Debug.LogError($"Failed to save job {job.jobId} to SessionState: {ex.Message}");
            }
        }
        
        private static void SaveActiveJobIds()
        {
            try
            {
                var jobIds = new List<string>();
                foreach (var jobId in activeJobs.Keys)
                {
                    jobIds.Add(jobId);
                }
                string jobIdsJson = JsonConvert.SerializeObject(jobIds);
                SessionState.SetString(ACTIVE_JOBS_KEY, jobIdsJson);
            }
            catch (Exception ex)
            {
                Debug.LogError($"Failed to save active job IDs: {ex.Message}");
            }
        }
        
        private static void RestoreJobsFromSessionState()
        {
            try
            {
                // Get the list of active job IDs
                string jobIdsJson = SessionState.GetString(ACTIVE_JOBS_KEY, null);
                if (string.IsNullOrEmpty(jobIdsJson))
                {
                    Debug.Log("[TriggerDomainReload] No active jobs to restore.");
                    return;
                }
                
                var jobIds = JsonConvert.DeserializeObject<List<string>>(jobIdsJson);
                var restoredCount = 0;
                var completedCount = 0;
                
                foreach (var jobId in jobIds)
                {
                    string jobJson = SessionState.GetString(JOBS_KEY_PREFIX + jobId, null);
                    if (!string.IsNullOrEmpty(jobJson))
                    {
                        try
                        {
                            var job = JsonConvert.DeserializeObject<DomainReloadJob>(jobJson);
                            if (job != null)
                            {
                                // Check if this job likely caused the domain reload
                                if (job.status == "running" && 
                                    (job.action == "domain_reload" || job.action == "compile_and_reload"))
                                {
                                    // Mark it as completed since we've successfully reloaded
                                    job.status = "completed";
                                    job.compilationSucceeded = true;
                                    job.message = "Domain reload completed successfully";
                                    job.compilationLogs.Add($"[{DateTime.Now:HH:mm:ss}] Domain reloaded - job completed");
                                    SaveJobToSessionState(job);
                                    completedCount++;
                                    Debug.Log($"[TriggerDomainReload] Job {job.jobId} marked as completed after domain reload");
                                }
                                else if (job.status == "running")
                                {
                                    // Other running jobs should be restored to continue processing
                                    lock (lockObj)
                                    {
                                        activeJobs[job.jobId] = job;
                                        restoredCount++;
                                    }
                                }
                            }
                        }
                        catch (Exception ex)
                        {
                            Debug.LogError($"Failed to restore job {jobId}: {ex.Message}");
                        }
                    }
                }
                
                Debug.Log($"[TriggerDomainReload] Restored {restoredCount} running jobs, marked {completedCount} as completed after domain reload.");
                
                // Update the active job IDs list
                SaveActiveJobIds();
            }
            catch (Exception ex)
            {
                Debug.LogError($"Failed to restore jobs from SessionState: {ex.Message}");
            }
        }
        
        private static void RegisterCompilationHandlers()
        {
            CompilationPipeline.compilationStarted += OnCompilationStarted;
            CompilationPipeline.compilationFinished += OnCompilationFinished;
            CompilationPipeline.assemblyCompilationFinished += OnAssemblyCompilationFinished;
            Application.logMessageReceived += OnLogMessageReceived;
        }
        
        private static void UnregisterCompilationHandlers()
        {
            CompilationPipeline.compilationStarted -= OnCompilationStarted;
            CompilationPipeline.compilationFinished -= OnCompilationFinished;
            CompilationPipeline.assemblyCompilationFinished -= OnAssemblyCompilationFinished;
            Application.logMessageReceived -= OnLogMessageReceived;
        }
        
        private static void OnCompilationStarted(object obj)
        {
            Debug.Log("[TriggerDomainReload] Compilation started");
            
            lock (lockObj)
            {
                foreach (var job in activeJobs.Values)
                {
                    if (job.status == "running")
                    {
                        job.compilationStarted = true;
                        job.compilationLogs.Add($"[{DateTime.Now:HH:mm:ss}] Compilation started");
                        SaveJobToSessionState(job);
                    }
                }
            }
        }
        
        private static void OnCompilationFinished(object obj)
        {
            Debug.Log("[TriggerDomainReload] Compilation finished");
            
            lock (lockObj)
            {
                foreach (var job in activeJobs.Values)
                {
                    if (job.status == "running")
                    {
                        job.compilationFinished = true;
                        job.compilationLogs.Add($"[{DateTime.Now:HH:mm:ss}] Compilation finished");
                        SaveJobToSessionState(job);
                    }
                }
            }
        }
        
        private static void OnAssemblyCompilationFinished(string assemblyPath, CompilerMessage[] messages)
        {
            lock (lockObj)
            {
                foreach (var job in activeJobs.Values)
                {
                    if (job.status == "running" && job.includeLogs)
                    {
                        job.compilationLogs.Add($"[{DateTime.Now:HH:mm:ss}] Assembly compiled: {assemblyPath}");
                        
                        foreach (var message in messages)
                        {
                            if (message.type == CompilerMessageType.Error)
                            {
                                job.compilationSucceeded = false;
                                job.compilationErrors.Add(new CompilationError
                                {
                                    type = message.type.ToString(),
                                    message = message.message,
                                    file = message.file,
                                    line = message.line,
                                    column = message.column
                                });
                                job.compilationLogs.Add($"[ERROR] {message.file}({message.line},{message.column}): {message.message}");
                            }
                            else if (message.type == CompilerMessageType.Warning)
                            {
                                job.compilationLogs.Add($"[WARNING] {message.file}({message.line},{message.column}): {message.message}");
                            }
                        }
                        
                        SaveJobToSessionState(job);
                    }
                }
            }
        }
        
        private static void OnLogMessageReceived(string logString, string stackTrace, LogType type)
        {
            if (type == LogType.Error || type == LogType.Exception)
            {
                lock (lockObj)
                {
                    foreach (var job in activeJobs.Values)
                    {
                        if (job.status == "running" && job.includeLogs)
                        {
                            job.compilationLogs.Add($"[{type}] {logString}");
                            if (!string.IsNullOrEmpty(stackTrace))
                            {
                                job.compilationLogs.Add($"Stack trace: {stackTrace}");
                            }
                            SaveJobToSessionState(job);
                        }
                    }
                }
            }
        }
        
        private static List<string> FilterLogs(List<string> logs, string logLevel)
        {
            if (logLevel == "all")
            {
                return new List<string>(logs);
            }
            
            var filtered = new List<string>();
            foreach (var log in logs)
            {
                bool include = false;
                
                switch (logLevel)
                {
                    case "error":
                        include = log.Contains("[ERROR]") || log.Contains("[Exception]");
                        break;
                    case "warning":
                        include = log.Contains("[ERROR]") || log.Contains("[Exception]") || log.Contains("[WARNING]");
                        break;
                }
                
                if (include)
                {
                    filtered.Add(log);
                }
            }
            
            return filtered;
        }
    }
}

File: UnityMcpServer/src/tools/trigger_domain_reload.py
"""Tool for triggering domain reload and script compilation in Unity."""

from typing import Dict, Any
import asyncio
import time
from mcp.server.fastmcp import FastMCP, Context
from unity_connection import get_unity_connection
import logging

logger = logging.getLogger(__name__)

def register_domain_reload_tools(mcp: FastMCP):
    """Register domain reload tool with the MCP server."""
    
    @mcp.tool()
    def trigger_domain_reload(
        ctx: Context,
        action: str = "compile_and_reload",
        wait_for_completion: bool = True,
        include_logs: bool = True,
        log_level: str = "all",
        timeout: float = 300,
        job_id: str = None
    ) -> Dict[str, Any]:
        """Triggers domain reload, script compilation, or asset refresh in Unity.
        
        Args:
            action: Operation to perform:
                    - 'refresh_assets': Refresh asset database
                    - 'compile_scripts': Request script compilation
                    - 'domain_reload': Force full domain reload
                    - 'compile_and_reload': Compile scripts then reload domain (default)
                    - 'get_status': Get status of a running job (requires job_id)
            wait_for_completion: Whether to wait for operation to complete (default: True)
                                When False, returns immediately with a job_id for polling
            include_logs: Include compilation logs in response (default: True)
            log_level: Minimum log level to include ('all', 'warning', 'error')
            timeout: Maximum seconds to wait for completion (default: 300)
            job_id: Job ID for status polling (only used with action='get_status')
            
        Returns:
            Dictionary with:
                - success: Whether the operation completed successfully
                - For new jobs: jobId, status='initiated', sessionId
                - For status checks: jobId, status, compilationErrors, logs, duration
        """
        try:
            connection = get_unity_connection()
            
            # Prepare command parameters
            params = {
                "action": action,
                "waitForCompletion": wait_for_completion,
                "includeLogs": include_logs,
                "logLevel": log_level,
                "timeout": timeout
            }
            
            # Add job_id for status polling
            if job_id is not None:
                params["jobId"] = job_id
            
            # Special case: if wait_for_completion is True and action is not get_status,
            # we need to implement our own polling loop
            if wait_for_completion and action != "get_status":
                logger.info(f"Triggering domain reload with action: {action} (synchronous)")
                
                # First, start the job
                params["waitForCompletion"] = False  # Force async on Unity side
                result = connection.send_command("trigger_domain_reload", params)
                
                if not result.get("success", False):
                    return {
                        "success": False,
                        "error": result.get("error", "Unknown error occurred"),
                        "message": result.get("message", "Failed to trigger domain reload")
                    }
                
                # Extract job ID from response
                data = result.get("data", {})
                job_id = data.get("jobId")
                
                if not job_id:
                    return {
                        "success": False,
                        "error": "No job ID returned from Unity",
                        "message": "Failed to start domain reload job"
                    }
                
                logger.info(f"Domain reload job started: {job_id}")
                
                # Now poll for completion
                poll_interval = 0.5  # Poll every 500ms
                elapsed = 0.0
                
                while elapsed < timeout:
                    # Wait before polling
                    time.sleep(poll_interval)
                    elapsed += poll_interval
                    
                    # Poll for status
                    status_params = {
                        "action": "get_status",
                        "jobId": job_id
                    }
                    
                    try:
                        status_result = connection.send_command("trigger_domain_reload", status_params)
                    except Exception as e:
                        # Connection might be lost during domain reload
                        logger.warning(f"Lost connection during polling (expected during domain reload): {e}")
                        time.sleep(2)  # Wait a bit longer before retrying
                        elapsed += 2
                        
                        # Try to reconnect
                        try:
                            connection = get_unity_connection()
                            continue
                        except:
                            logger.warning("Failed to reconnect, will keep trying...")
                            continue
                    
                    if not status_result.get("success", False):
                        # If we can't get status, it might be due to domain reload
                        logger.warning(f"Failed to get job status: {status_result.get('error', 'Unknown error')}")
                        continue
                    
                    status_data = status_result.get("data", {})
                    job_status = status_data.get("status", "unknown")
                    
                    # Check if job is complete
                    if job_status != "running":
                        # Job is done, return the full status
                        response = {
                            "success": status_data.get("compilationSucceeded", True),
                            "status": job_status,
                            "duration": status_data.get("duration", elapsed),
                            "message": status_data.get("message", "")
                        }
                        
                        # Add optional fields if present
                        if "compilationErrors" in status_data:
                            response["compilation_errors"] = status_data["compilationErrors"]
                        if "logs" in status_data:
                            response["logs"] = status_data["logs"]
                        if "wasCompiling" in status_data:
                            response["was_already_compiling"] = status_data["wasCompiling"]
                        if "hadToRefresh" in status_data:
                            response["had_to_refresh"] = status_data["hadToRefresh"]
                        
                        logger.info(f"Domain reload job {job_id} completed with status: {job_status}")
                        return response
                
                # Timeout reached
                return {
                    "success": False,
                    "status": "timeout",
                    "duration": elapsed,
                    "message": f"Operation timed out after {timeout} seconds",
                    "job_id": job_id
                }
                
            else:
                # Async mode or status polling - just pass through to Unity
                logger.info(f"Sending domain reload command: {action}")
                result = connection.send_command("trigger_domain_reload", params)
                
                if not result.get("success", False):
                    return {
                        "success": False,
                        "error": result.get("error", "Unknown error occurred"),
                        "message": result.get("message", "Failed to execute domain reload command")
                    }
                
                # Extract and format response data
                data = result.get("data", {})
                
                # For new jobs (async mode)
                if "jobId" in data and action != "get_status":
                    return {
                        "success": True,
                        "job_id": data.get("jobId"),
                        "status": data.get("status", "initiated"),
                        "message": data.get("message", "Domain reload job started"),
                        "session_id": data.get("sessionId"),
                        "was_already_compiling": data.get("wasCompiling", False),
                        "had_to_refresh": data.get("hadToRefresh", False)
                    }
                
                # For status checks
                else:
                    response = {
                        "success": data.get("compilationSucceeded", True),
                        "status": data.get("status", "unknown"),
                        "duration": data.get("duration", 0),
                        "message": data.get("message", "")
                    }
                    
                    # Add optional fields if present
                    if "jobId" in data:
                        response["job_id"] = data["jobId"]
                    if "compilationErrors" in data:
                        response["compilation_errors"] = data["compilationErrors"]
                    if "logs" in data:
                        response["logs"] = data["logs"]
                    if "wasCompiling" in data:
                        response["was_already_compiling"] = data["wasCompiling"]
                    if "hadToRefresh" in data:
                        response["had_to_refresh"] = data["hadToRefresh"]
                    if "sessionId" in data:
                        response["session_id"] = data["sessionId"]
                    
                    return response
            
        except ConnectionError as e:
            logger.error(f"Unity connection error: {e}")
            return {
                "success": False,
                "error": f"Unity connection error: {str(e)}",
                "message": "Failed to connect to Unity Editor"
            }
        except Exception as e:
            logger.error(f"Unexpected error in trigger_domain_reload: {e}")
            return {
                "success": False,
                "error": str(e),
                "message": "An unexpected error occurred"
            }